pylint/checkers/base.py

```diff
@@ -149,6 +153,7 @@ class CamelCaseStyle(NamingStyle):
 class PascalCaseStyle(NamingStyle):
     """Regex rules for PascalCase naming style."""
 
+    TYPE_VAR_RGX = re.compile(r"[^\W\da-z]+?_?([^\W\da-z_][^\WA-Z_]+)*(_co(ntra)?)?$")
```

Comment:
Just got a change to read what PEP8 had to say about it.
https://www.python.org/dev/peps/pep-0008/#type-variable-names

As I understand it, the pattern should be similar to normal class names, except for the `_co` and `_contra` suffixes.
So by default any of these should be allowed
```py
from typing import TypeVar

T = TypeVar("T")
AnyStr = TypeVar("AnyStr")
Num = TypeVar("Num")

VT_co = TypeVar('VT_co', covariant=True)
KT_contra = TypeVar('KT_contra', contravariant=True)
```

We should probably also allow a leading underscore
```py
_T = TypeVar("_T")
_Num = TypeVar("_Num")
```

Reply:
The first examples are all supported.

Why would we allow a leading underscore? The examples in the PEP and the PEP introducing `TypeVar` don't seem to use them.
Is this to declare them privately? If so, I'll have to check if that is not something taken care of by the checker at "check-time". The other patterns also don't seem to include leading underscores.

Reply:
`AnyStr` and `Num` won't work. The first part currently needs to be all upper case: `[^\W\da-z]+?`. This is basically equivalent to `[A-Z_]+?`. Thus a leading `_` is already handled.

> Why would we allow a leading underscore? The examples in the PEP and the PEP introducing TypeVar don't seem to use them. Is this to declare them privately?

Yes. We support it for other names, too.